probe-rs/src/architecture/arm/communication_interface.rs

```diff
@@ -164,31 +168,43 @@ fn get_debug_port_version(probe: &mut Probe) -> Result<DebugPortVersion, DebugPr
     Ok(DebugPortVersion::from(dpidr.version()))
 }
 
-impl InnerArmCommunicationInterface {
-    fn new(mut probe: Probe) -> Result<Self, (Probe, DebugProbeError)> {
-        // Check the version of debug port used
-        let version = match get_debug_port_version(&mut probe) {
-            Ok(version) => version,
-            Err(e) => return Err((probe, e)),
-        };
-
-        log::debug!("Debug Port version: {:?}", version);
-
-        let mut s = Self {
-            probe,
-            debug_port_version: version,
-            current_dpbanksel: 0,
-            current_apsel: 0,
-            current_apbanksel: 0,
-        };
+impl<'a> ArmCommunicationInterface<'a> {
+    pub fn new(
+        probe: &'a mut Probe,
+        state: &'a mut ArmCommunicationInterfaceState,
+    ) -> Result<Option<Self>, DebugProbeError> {
+        if probe.has_dap_interface() {
+            let mut s = Self { probe, state };
+
+            if !s.state.initialized() {
+                s.enter_debug_mode()?;
+                s.state.initialize();
+            }
 
-        if let Err(e) = s.enter_debug_mode() {
-            Err((s.probe, e.into()))
+            Ok(Some(s))
         } else {
-            Ok(s)
+            log::debug!("No DAP interface available on Probe");
+
+            Ok(None)
         }
     }
 
+    pub fn create_state(
+        probe: &mut Probe,
+    ) -> Result<ArmCommunicationInterfaceState, DebugProbeError> {
+        ArmCommunicationInterfaceState::new(probe)
+    }
+
+    pub fn borrow(&mut self) -> ArmCommunicationInterface<'_> {
```

Comment:
I don't understand the purpose of this method. Isn't `interface.borrow()` basically the same as `&mut interface`?